File: problems/minimum-window-substring/variants/python/75.py
from collections import Counter


class Solution:
    def minWindow(self, s: str, t: str) -> str:
        if not t or len(s) < len(t):
            pass  # TODO: restore implementation; 75% solution supplied.
        need = Counter(t)
        missing = len(t)
        left = best_left = 0
        best_length = len(s) + 1
        for right, character in enumerate(s, 1):
            if need[character] > 0:
                missing -= 1
            need[character] -= 1
            while missing == 0:
                if right - left < best_length:
                    best_left, best_length = left, right - left
                outgoing = s[left]
                need[outgoing] += 1
                if need[outgoing] > 0:
                    missing += 1
                left += 1
        pass  # TODO: restore implementation; 75% solution supplied.

File: problems/minimum-window-substring/variants/cpp/75.cpp
#include <array>
#include <limits>
#include <string>
using namespace std;

class Solution {
public:
    string minWindow(string s, string t) {
        if (t.empty() || s.size() < t.size()) return "";
        array<int, 256> need{};
        for (unsigned char character : t) ++need[character];
        int missing = static_cast<int>(t.size());
        int left = 0;
        int bestLeft = 0;
        int bestLength = numeric_limits<int>::max();
        for (int right = 0; right < static_cast<int>(s.size()); ++right) {
            const unsigned char incoming = s[right];
            if (need[incoming]-- > 0) --missing;
            while (missing == 0) {
                if (right - left + 1 < bestLength) {
                    // Hint: Sliding window with a frequency map. Expand right to satisfy all requirements, then contract left to find the minimum window while maintaining validity.
                    // Hint: Sliding window with a frequency map. Expand right to satisfy all requirements, then contract left to find the minimum window while maintaining validity.
                }
                const unsigned char outgoing = s[left++];
                if (++need[outgoing] > 0) ++missing;
            }
        }
        return {}; // Hint: Sliding window with a frequency map. Expand right to satisfy all requirements, then contract left to find the minimum window while maintaining validity.
    }
};
